cyten/models/degrees_of_freedom.py

```diff
@@ -0,0 +1,759 @@
+"""Defines classes describing the local physical Hilbert spaces.
+
+The :class:`DegreeOfFreedom` is the prototype, read its docstring.
+All other classes are base classes from which sites are derived.
+"""
+# Copyright (C) TeNPy Developers, Apache license
+from __future__ import annotations
+import numpy as np
+from typing import Literal, Sequence
+from functools import reduce
+from math import comb
+
+from ..backends import TensorBackend, get_backend
+from ..backends.abstract_backend import Block
+from ..spaces import ElementarySpace
+from ..tensors import DiagonalTensor, SymmetricTensor
+from ..symmetries import (
+    FermionNumber, FermionParity, U1Symmetry, ZNSymmetry, SU2Symmetry, Sector, Symmetry,
+    NoSymmetry, ProductSymmetry, BraidingStyle
+)
+
+
+class DegreeOfFreedom:
+    """Collects necessary information about a local degree of freedom of a lattice model.
+
+    A degree of freedom defines the local Hilbert space in terms of its :attr:`leg`.
+    This involves a choice for the local basis.
+    Moreover, it exposes the symmetric single-site operators.
+    Multi-site operators, on the other hand, are represented by :class:`Coupling` s.
+
+    .. todo ::
+        How to handle JW strings?
+
+    .. todo ::
+        Mechanism to change the symmetry
+
+    Attributes
+    ----------
+    leg : ElementarySpace
+        The local physical Hilbert space.
+    state_labels : {str: int}
+        Optional labels for the local basis states. Any state may have multiple labels, or none.
+    onsite_operators : {str: SymmetricTensor}
+        The available on-site operators. Note: which operators are available typically depends
+        on what symmetry is enforced. Operators that are symmetric under a small symmetry may
+        not be symmetric under a larger symmetry, and are thus not available as `onsite_operators`.
+        Each must have the :attr:`leg` of the site as the only factor in its domain and codomain.
+
+    Examples
+    --------
+    TODO
+    """
+
+    def __init__(self, leg: ElementarySpace, state_labels: dict[str, int] = None,
+                 onsite_operators: dict[str, SymmetricTensor] = None,
+                 backend: TensorBackend = None, default_device: str = None):
+        self.leg = leg
+        if state_labels is None:
+            state_labels = {}
+        self.state_labels = state_labels
+        if backend is None:
+            backend = get_backend(symmetry=leg.symmetry)
+        self.backend = backend
+        if default_device is None:
+            default_device = 'cpu'
+        self.default_device = default_device
```

Comment:
long term todo to clean up handling of devices and dtypes, in particular where defaults come from